workload/tpcc: initial data load perf improvements

danhhz opened the pull request:
See individual commits for details

Touches #34809

danhhz committed:
workload/tpcc: reduce some allocations

Use bufalloc.ByteAllocator in randStringFromAlphabet instead of
allocating many small byte arrays.

Benchmark results:
- InitTPCC only measures generating initial data
- ImportFixtureTPCC measures loading initial data with
  COCKROACH_IMPORT_WORKLOAD_FASTER=true, which skips the CSV roundtrip.

    name                     old time/op    new time/op    delta
    InitTPCC/warehouses=1-8     2.33s ± 0%     2.25s ± 1%   -3.04%  (p=0.008 n=5+5)
    ImportFixtureTPCC-8         5.00s ± 2%     4.83s ± 2%   -3.32%  (p=0.008 n=5+5)

    name                     old speed      new speed      delta
    InitTPCC/warehouses=1-8  30.4MB/s ± 0%  31.4MB/s ± 1%   +3.14%  (p=0.008 n=5+5)
    ImportFixtureTPCC-8      17.7MB/s ± 2%  18.3MB/s ± 2%   +3.40%  (p=0.008 n=5+5)

    name                     old alloc/op   new alloc/op   delta
    InitTPCC/warehouses=1-8     329MB ± 0%     246MB ± 0%  -25.28%  (p=0.008 n=5+5)
    ImportFixtureTPCC-8        3.70GB ± 0%    3.61GB ± 0%   -2.20%  (p=0.008 n=5+5)

    name                     old allocs/op  new allocs/op  delta
    InitTPCC/warehouses=1-8     9.38M ± 0%     5.60M ± 0%  -40.27%  (p=0.008 n=5+5)
    ImportFixtureTPCC-8         36.0M ± 0%     32.2M ± 0%  -10.49%  (p=0.008 n=5+5)

Touches #34809

Release note: None

danhhz committed:
workload/tpcc: use a faster random number generator

Background: https://github.com/golang/go/issues/21835

The current thinking for go2 is to switch to a Permuted Congruential
Generator (PCG), which is much faster than the current "math/rand"
generator. It's smaller in memory (128-bit vs 607 64-bit words) and
_much_ faster to seed (which is a common operation in workload, which
uses seeding to make deterministic results).

The linked issues claims "The current implementation, in pure Go, is
somewhat slower than math/rand, but could be comparably fast or even
faster given compiler support for the 128-bit multiply inside." but for
our needs it seems to already be faster.

Benchmark results:
- InitTPCC only measures generating initial data
- ImportFixtureTPCC measures loading initial data with
  COCKROACH_IMPORT_WORKLOAD_FASTER=true, which skips the CSV roundtrip.

    name                     old time/op    new time/op    delta
    InitTPCC/warehouses=1-8     2.25s ± 1%     2.06s ± 1%  -8.07%  (p=0.008 n=5+5)
    ImportFixtureTPCC-8         4.90s ± 3%     4.64s ± 2%  -5.44%  (p=0.008 n=5+5)

    name                     old speed      new speed      delta
    InitTPCC/warehouses=1-8  31.5MB/s ± 1%  34.3MB/s ± 1%  +8.80%  (p=0.008 n=5+5)
    ImportFixtureTPCC-8      18.0MB/s ± 3%  19.1MB/s ± 2%  +5.77%  (p=0.008 n=5+5)

    name                     old alloc/op   new alloc/op   delta
    InitTPCC/warehouses=1-8     246MB ± 0%     246MB ± 0%    ~     (p=0.095 n=5+5)
    ImportFixtureTPCC-8        3.61GB ± 0%    3.61GB ± 0%    ~     (p=0.548 n=5+5)

    name                     old allocs/op  new allocs/op  delta
    InitTPCC/warehouses=1-8     5.60M ± 0%     5.60M ± 0%  +0.06%  (p=0.008 n=5+5)
    ImportFixtureTPCC-8         32.2M ± 0%     32.3M ± 0%  +0.08%  (p=0.008 n=5+5)

Touches #34809

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35322)
<!-- Reviewable:end -->

danhhz committed:
workload/tpcc: faster random string generation

The slowest part of generating random strings is generating the random
number. Previously, we made a uint64 and threw away most of it to pick
one character from an alphabet. Instead, pick as many characters as we
can get from 64 bits: floor(log(math.MaxUint64)/log(len(alpabet)).

New random string generation microbenchmark:

    name                      time/op
    RandStringFast/letters-8    75.3ns ± 2%
    RandStringFast/numbers-8    74.2ns ± 1%
    RandStringFast/aChars-8     84.5ns ± 1%

    name                      speed
    RandStringFast/letters-8   345MB/s ± 2%
    RandStringFast/numbers-8   351MB/s ± 1%
    RandStringFast/aChars-8    308MB/s ± 1%

Diffs in TPCC data generation and TPCC IMPORT:

    name                     old time/op    new time/op     delta
    InitTPCC/warehouses=1-8     2.03s ± 2%      0.65s ± 1%   -67.85%  (p=0.008 n=5+5)
    ImportFixtureTPCC-8         4.11s ± 5%      3.75s ± 2%    -8.67%  (p=0.008 n=5+5)

    name                     old speed      new speed       delta
    InitTPCC/warehouses=1-8  34.9MB/s ± 2%  108.5MB/s ± 1%  +211.01%  (p=0.008 n=5+5)
    ImportFixtureTPCC-8      21.5MB/s ± 5%   23.6MB/s ± 2%    +9.45%  (p=0.008 n=5+5)

Touches #34809

Release note: None

danhhz commented:
Nathan's out this week, removing him as a possible reviewer

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, 12 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35322#-:-L_8SeeWCOljGkL5Q1XW:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @petermattis)



<!-- Sent from Reviewable.io -->

danhhz commented:
Thanks for the review!

bors r=petermattis
